compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1571,24 +1649,95 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
     fn report_similar_impl_candidates(
         &self,
         impl_candidates: Vec<ImplCandidate<'tcx>>,
+        trait_ref: ty::PolyTraitRef<'tcx>,
         err: &mut Diagnostic,
-    ) {
+    ) -> bool {
+        let report = |mut candidates: Vec<TraitRef<'_>>, err: &mut Diagnostic| {
+            candidates.sort();
+            candidates.dedup();
+            let len = candidates.len();
+            if candidates.len() == 0 {
+                return false;
+            }
+            if candidates.len() == 1 {
+                err.highlighted_help(vec![
+                    (
+                        format!("the trait `{}` ", candidates[0].print_only_trait_path()),
+                        Style::NoStyle,
+                    ),
+                    ("is".to_string(), Style::Highlight),
+                    (" implemented for `".to_string(), Style::NoStyle),
+                    (candidates[0].self_ty().to_string(), Style::Highlight),
+                    ("`".to_string(), Style::NoStyle),
+                ]);
+                return true;
+            }
+            let trait_ref = TraitRef::identity(self.tcx, candidates[0].def_id);
+            // Check if the trait is the same in all cases. If so, we'll only show the type.
+            // FIXME: there *has* to be a better way!
+            let mut traits: Vec<_> = candidates
+                .iter()
+                .map(|c| format!("{}", c).split(" as ").last().unwrap().to_string())
+                .collect();
```

Comment:
Calling this out: it's a hack and I was too sleepy to come up with a replacement.

I think `print_only_trait_path` (as opposed to `print_only_trait_name`) will work?